### PR TITLE
ICU-22027 Add Temporal related Calendar API 

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22027
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [X] API docs and/or User Guide docs changed or added, if applicable

API proposal 
https://docs.google.com/document/d/1UYriEzzExiLhi2RD3zjTsI5UQHv1dXaFqrct7yXNdCA/edit#heading=h.x9obor85vpx9

Design Doc https://docs.google.com/document/d/15ViyC9s0k3VEDwBmAkKxxz4IadZ6QrAIoETkdkF0cVA/edit#

[ICU-22027]: https://unicode-org.atlassian.net/browse/ICU-22027?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ICU-1234]: https://unicode-org.atlassian.net/browse/ICU-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ICU-1234]: https://unicode-org.atlassian.net/browse/ICU-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by FrankYFTang** [comment]
sorry, please hold the review, there is still some bug in this PR

**Comment by FrankYFTang** [comment]
I have no idea why "adaboost-icu4j-build-and-test (pull_request)" is broken here since I didn't touch any Java code in this PR

**Review by richgillam — APPROVED** [review]
I got about a third of the way through this before I realized that I'm not really qualified to review this code, but for what it's worth, the structural stuff all makes sense to me and I have no reason to doubt how you're doing the calendrical calculations (although I'd really like another pair of eyes on all this stuff).  I did have a few small questions about the documentation, but nothing too earth-shattering.

I didn't actually review the unit tests in detail-- I got lost very quickly.  I apologize for this.

File: icu4c/source/i18n/calendar.cpp
Comment (FrankYFTang) [reply icu4c/source/i18n/calendar.cpp:2795]: Right, currently, the only subclass of Calendar will set UCAL_IS_LEAP_MONTH is ChineseCalendar which is also a superclass of DangiCalendar. So only ChineseCalendar and DangiCalendar will set that bit. And we handle that in the ChineseCalendar::getTemporalMonthCode but want to ensure in the future if some other subclass of Calendar also set UCAL_IS_LEAP_MONTH we will bark here and force them to add an override of  getTemporalMonthCode()

File: icu4c/source/i18n/chnsecal.h
Comment (FrankYFTang) [reply icu4c/source/i18n/chnsecal.h:133]: What I want to say is for non leap year, the return values for that year are "M01" .. "M12". And for any given leap year, the return values for that year are "M01" .. "M12" and ONE AND ONLY ONE of a value in "M01L" .. "M12L" for that particular leap year. (but it won't have two different one in that same year, no two leap months in a leap year).

File: icu4c/source/i18n/calendar.cpp
Comment (richgillam) [reply icu4c/source/i18n/calendar.cpp:2795]: Thanks; I eventually figured that out after reading further into your PR.  It might be worth a quick note here explaining what you're doing (then again, it might not).

File: icu4c/source/i18n/chnsecal.h
Comment (richgillam) [reply icu4c/source/i18n/chnsecal.h:133]: What you said here was very clear; maybe that's how you should word it in the comments.  :-)

File: icu4c/source/i18n/chnsecal.h
Comment (FrankYFTang) [reply icu4c/source/i18n/chnsecal.h:203]: so the issue is in Chinese calendar, the calculation of leap is actually not determined by a 'year' (Chinese Jan-1 to Dec-30) but determined by a Sui (between two Winter Solstices) . The variable used to call isLeapYear which is misleading because that boolean value is used to indicate there is a leap month in the current "sui" (but the begin/end of a "sui" is not the begin/end of a "year") but not necessary in the current "year".

Notice "sui" is not a Chinese "year". So we have three (not two) different concepts
1) a Gregorian Year
2) a Chinese Year (12 or 13 months)- which go with the Moon.
3) a Sui (the period between two Winter Solstices,  Winter Solstices are usually on Gregorian Dec 21 or Dec 22, and go with the Sun, not Moon)

That variable is indicating the particular moment (the date value) is in a Sui which has a leap month. We could name that variable inLeapSui but that is not the same as inLeapYear.

The ECMA Temporal API asked us to return the answer of it is in a leap YEAR or not, not in a leap Sui or not, so we cannot just return that value. 

There are dates in a Leap year which is NOT in a leap Sui. For example, If we have a leap April in Chinese year, any date in the next year is always not in a leap year (since there are not consective leap year in Chinese calendar), and any dates after the Winter Solstices (Gregorian Calendar Dec 21 or 22, always a date in Chinese calendar the "11th month" since Chinese calendar guarantee Winter Solstices fall into Chinese "11th month" ) of that year will be still in the Leap Year, but not in the leap Sui. Similarly, For the date in the previous year after the Winter Solstices, that date (in the year prior to the one have Leap April) will be in non leap year but that variable will be true because there is a leap month between that two Winter Solstices.

Comment (richgillam) [reply icu4c/source/i18n/chnsecal.h:203]: Thank you for that explanation; that was very clear and answers my questions.

To make sure I understand, what you're saying is that a calendar year in the Chinese year straddles two "sui"s, either of which (but not both) may contain a leap month.  If either does, the calendar year can be considered a "leap year."  (Or does the leap month itself have to be within that calendar year?  And is that the Chinese-calendar year the "is leap year" stuff is talking about, or the corresponding Gregorian year?)  The `hasLeapMonthBetweenWinterSolstices()` method tells you whether the "sui" containing the current date has a leap month, but if it returns false, the calendar year might still be a "leap year" because the _other_ "sui" might contain a leap month.

Do I have this right?

Your explanation is great, but long.  I wonder if it's possible to condense it a little and put that into the source code somewhere?  Alternatively, there might be a better name for the function, but I can't think of one.

File: icu4c/source/i18n/unicode/calendar.h
Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h:1375]: I am not 100% sure. Per ref in https://unicode-org.atlassian.net/browse/CLDR-11440 

http://www2.math.uu.se/~svante/papers/calendars/tibet.pdf 

"The regular months in each year are numbered 1–12. A leap month takes
the same number (and belongs to the same year) as the following month (as in Indian calendars [2]).14 In a leap year there are thus two months having the same number and, by the rule just given, the first of these is the leap month." 

It seems that ref suggest the Leap Jan in Tibetan calendar, happen before Jan should " first of these is the leap month." but "takes the same number"

so they may prefer
"M01L", "M01", "M02"....

I was aware of such leap month scheme (leap before the Jan) so I spec with M00L in mind but not I am not sure eventually they will use "M00L" or "M01L" for that leap month before Jan.

File: icu4c/source/i18n/unicode/calendar.h
Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h:1402]: No, 
"
For Coptic and Ethiopian calendar, the Temporal monthCode values for any years are "M01" to "M13".
"
AND
"
In a calendar with a leap month at the start of some years, the month code of that month would be "M00L".
"

NOT in Coptic and Ethiopian calendar.

File: icu4c/source/i18n/chnsecal.h
Comment (FrankYFTang) [reply icu4c/source/i18n/chnsecal.h:203]: "If either does, the calendar year can be considered a "leap year.""

NO. the leap year cannot be derived from the leap sui. Look at the following
```
-----WS---G----C--------------------------------WS---G----C-------------------
-----|-------------------Sui--------------------|-----------------------------
-Non Leap Year-|--------------Leap Year-------------------|-Non Leap Year-----
1      2                 3                           4           5
```
where 
* WS is the Winter Solstice
* G is the start of a Gregoria Year
* C is the start of a Chinese year 

A date is in a leap year if there is a leap month between two Cs, 
That variable indicate there is a leap month between two WSs.
so the condiction is not just "If either does". It is acutally checking two different thing. 
The leap month may happen in point 3 (from the beginning of the year till WS) or point 4 (between WS to the end of the year). It is very rare for it to happen in point 4 but still possible.
That varialbe only indicate there is a leap month fall into the current Sui. 
If the date value is in 1 or 2, (but the leap month is in point 3 or 4), it is NOT in a leap year
For example, if a leap month fall into point 3, then point 2 and point 3 are both in the same Sui, but point 2 is not in a leap year but point 3 is in a leap year. and the value of that variable is true for both point 2 and point 3. 

Also, for some rarely happen year (for example 2033), the leap month may happen in point 4, in that case, both point 3 and point 4 are in leap year but not point 1 , 2 nor 5. but for that variable (hasLeapMonthBetweenWinterSolstices) , it is true in point 4 and 5 but false in point 1, 2, and 3.

Comment (richgillam) [reply icu4c/source/i18n/chnsecal.h:203]: Okay, that all makes sense.  Thanks again.

File: icu4c/source/i18n/unicode/calendar.h
Comment (richgillam) [reply icu4c/source/i18n/unicode/calendar.h:1375]: It does kind of sound that way, but at least I understand your intent, and I have no objections to that.

File: icu4c/source/i18n/unicode/calendar.h
Comment (richgillam) [reply icu4c/source/i18n/unicode/calendar.h:1402]: Okay.  In that case, I'd recommend replacing "In a calendar with a leap month at the start..." to "For the XX and YY calendars, which add leap months at the beginning of the year...", where XX and YY are the actual calendars that do that.  Or at least start the sentence with "for" to maintain the parallelism and clarify that you're not talking about the Coptic and Ethiopic calendars.

Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h:1402]: This comment in this header is part of the API proposal we discussed in ICUTC. I probably need to bring that back to ICU TC to get everyone agree upon first. 
How aboout 

```
For the calendar which the leap month at the start of some years, the month code of that month would be "M00L".
```

I need to talk to Peter about this. because after reading http://www2.math.uu.se/~svante/papers/calendars/tibet.pdf it seems they prefer that leap month coded as "M01L" instead of "M00L" and just want the leap month happen before the month with the same number. @pedberg-icu

Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h:1402]: I wonder should I just remove all the "M00L" comment all together in the API and wait till the picture is clear about Tibeten calendar first.

Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h:1402]: see http://www2.math.uu.se/~svante/papers/calendars/tibet.pdf

"Since these correspondences are not perfect, there are occasionally two months with the same number, in which case the first of them is regarded as a leap month"

File: icu4c/source/i18n/chnsecal.h
Comment (FrankYFTang) [reply icu4c/source/i18n/chnsecal.h:133]: remove that part after approved by ICU TC.

File: icu4c/source/i18n/unicode/calendar.h
Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h]: fixed

File: icu4c/source/i18n/unicode/calendar.h
Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h]: Removed that part after approved by ICUTC yesterday.

File: icu4c/source/i18n/unicode/calendar.h
Comment (FrankYFTang) [reply icu4c/source/i18n/unicode/calendar.h]: removed that part after approved by ICU TC

**Comment by FrankYFTang** [comment]
Please take another look

**Review by richgillam — DISMISSED** [review]
Still LGTM.

**Comment by FrankYFTang** [comment]
Rich- I squeshed PTAL
Peter- any feedback ?

**Review by richgillam — APPROVED** [review]
Still LGTM.

**Comment by FrankYFTang** [comment]
Peter, if you have other concerns, please file bug against me and I will fix it in subsequent PR. Thanks

**Pull request merged**
